qiskit/algorithms/amplitude_amplifiers/amplification_problem.py

```diff
@@ -0,0 +1,198 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Amplification problem class."""
+
+from typing import Optional, Callable, Any, Union, List
+
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.library import GroverOperator
+from qiskit.quantum_info import Statevector
+
+
+class AmplificationProblem:
+    """The amplification problem is the input to amplitude amplification algorithms, like Grover.
+
+    This class contains all problem-specific information required to run an amplitude amplification
+    algorithm. It minimally contains the Grover operator. It can further hold some post processing
+    on the optimal bitstring.
+    """
+
+    def __init__(self,
+                 oracle: QuantumCircuit,
+                 state_preparation: Optional[QuantumCircuit] = None,
+                 grover_operator: Optional[QuantumCircuit] = None,
+                 post_processing: Optional[Callable[[str], Any]] = None,
+                 objective_qubits: Optional[Union[int, List[int]]] = None,
+                 is_good_state: Optional[Callable[[str], bool]] = None,
+                 ) -> None:
+        r"""
+        Args:
+            oracle: The oracle reflecting about the bad states.
+            state_preparation: A circuit preparing the input state, referred to as
+                :math:`\mathcal{A}`.
+            grover_operator: The Grover operator :math:`\mathcal{Q}` used as unitary in the
+                phase estimation circuit.
+            post_processing: A mapping applied to the most likely bitstring.
+            objective_qubits: If set, specifies the indices of the qubits that should be measured.
+                If None, all qubits will be measured. The ``is_good_state`` function will be
+                applied on the measurement outcome of these qubits.
+            is_good_state: A function to check whether a string represents a good state.
+        """
+        self._oracle = oracle
+        self._state_preparation = state_preparation
+        self._grover_operator = grover_operator
+        self._post_processing = post_processing
+        self._objective_qubits = objective_qubits
+        self._is_good_state = is_good_state
+
+    @property
+    def oracle(self) -> QuantumCircuit:
+        """Return the oracle.
+
+        Returns:
+            The oracle.
+        """
+        return self._oracle
+
+    @oracle.setter
+    def oracle(self, oracle: QuantumCircuit) -> None:
+        """Set the oracle.
+
+        Args:
+            oracle: The oracle.
+        """
+        self._oracle = oracle
+
+    @property
+    def state_preparation(self) -> QuantumCircuit:
+        r"""Get the state preperation operator :math:`\mathcal{A}`.
+
+        Returns:
+            The :math:`\mathcal{A}` operator as `QuantumCircuit`.
+        """
+        if self._state_preparation is None:
```

Comment:
This, unlike setters we have in other places, if it was set None does not return None back rather it returns a default object built instead of None. This is different, I think, than we do elsewhere where you get None back and can see directly you are set to default behavior. I imagine returning a valid object here is so it can be directly used with the caller dealing with None, where the caller in the case would be an algorithm versus the end user who may directly set/get here too.

Reply:
Yes this is to simplify calls to the property in the algorithms. We have it both ways e.g. the amplitude estimation algorithms also always return a valid object.

Reply:
Ok, maybe it adds too much complication for a method that the algorithm would use, which behaves as the getter you have, and a getter which returns the same thing that is set so None (default being set) is easily knowable by an end user/program using the problem. Maybe they do not have to know - I guess I did not notice with the Amplitude Estimation.